va/validation-authority.go

```diff
@@ -335,67 +337,100 @@ func (va *ValidationAuthorityImpl) fetchHTTP(identifier core.AcmeIdentifier, pat
+	// Make connections with SNI = nonceName
 	conn, err := tls.DialWithDialer(&net.Dialer{Timeout: validationTimeout}, "tcp", hostPort, &tls.Config{
 		ServerName:         zName,
 		InsecureSkipVerify: true,
 	})
 
+	va.log.Notice(fmt.Sprintf("%s [%s] Attempting to validate for %s %s", challenge.Type, identifier, hostPort, zName))
 	if err != nil {
-		challenge.Status = core.StatusInvalid
 		challenge.Error = &core.ProblemDetails{
 			Type:   parseHTTPConnError(err),
 			Detail: "Failed to connect to host for DVSNI challenge",
 		}
 		va.log.Debug(fmt.Sprintf("%s [%s] TLS Connection failure: %s", challenge.Type, identifier, err))
-		return challenge, err
+		return challenge.Error
 	}
 	defer conn.Close()
 
-	// Check that zName is a dNSName SAN in the server's certificate
+	// Check that zName is the sole dNSName SAN in the server's certificate
 	certs := conn.ConnectionState().PeerCertificates
 	if len(certs) == 0 {
 		challenge.Error = &core.ProblemDetails{
 			Type:   core.UnauthorizedProblem,
 			Detail: "No certs presented for TLS SNI challenge",
 		}
-		challenge.Status = core.StatusInvalid
-		return challenge, challenge.Error
+		return challenge.Error
 	}
-	for _, name := range certs[0].DNSNames {
-		if subtle.ConstantTimeCompare([]byte(name), []byte(zName)) == 1 {
-			challenge.Status = core.StatusValid
-			return challenge, nil
-		}
+	// Do not inspect all Subject Alternative Names, because otherwise
+	// a single certificate could contain all of the Zi values
+	name := certs[0].DNSNames[0]
+	if subtle.ConstantTimeCompare([]byte(name), []byte(zName)) == 1 {
+		return nil
 	}
 
 	challenge.Error = &core.ProblemDetails{
 		Type:   core.UnauthorizedProblem,
 		Detail: "Correct zName not found for TLS SNI challenge",
 	}
-	challenge.Status = core.StatusInvalid
-	return challenge, challenge.Error
+	return challenge.Error
+}
+
+func (va *ValidationAuthorityImpl) validateTLSWithZNames(identifier core.AcmeIdentifier, input core.Challenge, zNames []string) (core.Challenge, error) {
+	challenge := input
+
+	addr, allAddrs, problem := va.getAddr(identifier.Value)
+	// Right now we make all connections to one of the addresses; maybe we should do a different one each time?
+	challenge.ValidationRecord = []core.ValidationRecord{
+		core.ValidationRecord{
+			Hostname:          identifier.Value,
+			AddressesResolved: allAddrs,
+			AddressUsed:       addr,
+		},
+	}
+	if problem != nil {
+		challenge.Status = core.StatusInvalid
+		challenge.Error = problem
+		return challenge, challenge.Error
+	}
+	portString := strconv.Itoa(va.tlsPort)
+	hostPort := net.JoinHostPort(addr.String(), portString)
+	challenge.ValidationRecord[0].Port = portString
+
+	va.log.Notice(fmt.Sprintf("%s [%s] Preparing to validate for %#v", challenge.Type, identifier, zNames))
+	if len(zNames) > 1 {
+		for m := 0; m < 5; m++ {
+			// Randomly pick one of the N values Zi for each validation
+			j, err := rand.Int(rand.Reader, big.NewInt(challenge.N))
```

Comment:
You need to make sure that you don't repeat values here.  Probably makes sense to do something like this in advance of the loop:

```
validationValues := map[int64]bool{}
bigN := big.NewInt(challenge.N)
for len(validationValues) < challenge.RequiredValues {
  newRandomValue := rand.Int(rand.Reader, bigN)
  validationValues[newRandomValue.Int64()] = true
}
```